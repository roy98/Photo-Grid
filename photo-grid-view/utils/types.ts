export type PhotoUrl = {
	full?: string
	regular?: string
	thumbnail?: string
}

export type PhotoUser = {
	name: string
	portfolio_url: string
}

export type LightboxProps = {
	isOpen: boolean
	urls: PhotoUrl
	description?: string
	downloads?: number
	likes?: number
	user?: PhotoUser
	handleClose: () => void
}

export type SinglePhotoProps = {
	src: PhotoUrl
	handleOpenLightbox: () => void
}
