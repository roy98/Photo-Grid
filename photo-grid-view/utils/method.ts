import { Fetcher } from 'swr'

const getPhotoFromUnsplash = async (url: string) => {
	const res = await fetch(
		`${process.env.NEXT_PUBLIC_UNSPLASH_BASE_URL}${url}&client_id=${process.env.NEXT_PUBLIC_UNSPLASH_ACCESS_KEY}`
	)
	return res.json()
}

// Our Custom API Endpoint
const getPhotoFromPhotoBackendAPI = async (url: string) => {
	const res = await fetch(`${process.env.NEXT_PUBLIC_PHOTO_GRID_API}api${url}`)

	const transformedPhotos = await res.json()

	return transformedPhotos?.map((photo: any) => {
		// parse urls and user object
		photo.urls = JSON.parse(photo.urls)
		photo.user = JSON.parse(photo.user)
		return photo
	})
}

// SWR: a function that accepts the index and the previous page data, returns the key of a page
const getKey = (pageIndex, previousPageData) => {
	if (previousPageData && !previousPageData.length) return null // reached the end
	return `/photos?page=${pageIndex + 1}&per_page=10` // SWR key
}

// Wrap native fetch api for SWR
// Change the called method to either use Unsplash or our Custom API
const fetcher: Fetcher<string[]> = (...args: any) => getPhotoFromPhotoBackendAPI(args)

export { getPhotoFromUnsplash, getKey, fetcher }
