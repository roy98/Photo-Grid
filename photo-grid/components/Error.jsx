import Image from 'next/image'
import styles from '../styles/common.module.css'

export default function Error() {
	return (
		<div className={styles.errorContainer}>
			<Image src={'/images/error.svg'} width={'400px'} height={'400px'} />
			<p>An error occured. Kyndly refresh the page or retry later</p>
			<button>Reload</button>
		</div>
	)
}
