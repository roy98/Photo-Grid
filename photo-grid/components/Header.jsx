import styles from '../styles/common.module.css'

export default function Header() {
	return (
		<header className={styles.header}>
			<span>Photo Grid</span>
		</header>
	)
}
