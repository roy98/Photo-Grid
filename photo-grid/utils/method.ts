import { Fetcher } from 'swr'

const getPhotoFromUnsplash = async (url: string) => {
	const res = await fetch(
		`${process.env.NEXT_PUBLIC_UNSPLASH_BASE_URL}${url}&client_id=${process.env.NEXT_PUBLIC_UNSPLASH_ACCESS_KEY}`
	)
	return res.json()
}

// SWR: a function that accepts the index and the previous page data, returns the key of a page
const getKey = (pageIndex, previousPageData) => {
	if (previousPageData && !previousPageData.length) return null // reached the end
	return `/photos?page=${pageIndex + 1}&per_page=10` // SWR key
}

// Wrap native fetch api for SWR
const fetcher: Fetcher<string[]> = (...args: any) => getPhotoFromUnsplash(args)

export { getPhotoFromUnsplash, getKey, fetcher }
