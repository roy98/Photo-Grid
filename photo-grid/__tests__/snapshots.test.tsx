import { render } from '@testing-library/react'
import Header from '@/components/Header'
import Error from '@/components/Error'
import SinglePhoto from '@/components/SingleImage'
import { PhotoUrl } from '@/utils/types'

describe('Header', () => {
	it('renders header unchanged', () => {
		const { container } = render(<Header />)
		expect(container).toMatchSnapshot()
	})
})

describe('Error', () => {
	it('renders error component', () => {
		const { container } = render(<Error />)
		expect(container).toMatchSnapshot()
	})
})

describe('SinglePhoto', () => {
	const src: PhotoUrl = { regular: '/images/error.svg' }
	it('renders header unchanged', () => {
		const { container } = render(<SinglePhoto src={src} handleOpenLightbox={() => {}} />)
		expect(container).toMatchSnapshot()
	})
})
