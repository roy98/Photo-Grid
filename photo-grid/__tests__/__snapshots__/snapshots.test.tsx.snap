// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error renders error component 1`] = `
<div>
  <div
    class="errorContainer"
  >
    <span
      style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
    >
      <span
        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
      >
        <img
          alt=""
          aria-hidden="true"
          src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27400%27%20height=%27400%27/%3e"
          style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
        />
      </span>
      <img
        data-nimg="intrinsic"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
      <noscript />
    </span>
    <p>
      An error occured. Kyndly refresh the page or retry later
    </p>
    <button>
      Reload
    </button>
  </div>
</div>
`;

exports[`Header renders header unchanged 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="logoContainer"
    >
      <span
        class="appName"
      >
        Photo Grid
      </span>
      <span
        class="logo"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2750%27%20height=%2750%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="Vercel Logo"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </span>
    </div>
    <p>
      Powered by 
      <a
        href=""
      >
        Unsplash
      </a>
    </p>
  </header>
  <p
    style="padding: 3rem 2rem 2rem 2rem; text-align: center;"
  >
    The most beautiful royalty-free photos and images shared for free by talented creators.
  </p>
</div>
`;

exports[`SinglePhoto renders header unchanged 1`] = `
<div>
  <span
    style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
  >
    <span
      style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
    >
      <img
        alt=""
        aria-hidden="true"
        src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27400%27%20height=%27400%27/%3e"
        style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
      />
    </span>
    <img
      data-nimg="intrinsic"
      data-testid="single-image"
      decoding="async"
      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
    />
    <noscript />
  </span>
</div>
`;
